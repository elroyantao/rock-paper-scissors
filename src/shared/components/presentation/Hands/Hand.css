.Hand {
  width: 150px;
  max-width: 19%;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 80px;
}

.Hand--progress {
  composes: Hand;
  max-width: initial;
}

.Hand-icon {
  width: 50px;
  height: 50px;
  transition: width .5s, height .5s;
  border: solid 1px black;
  border-radius: 75px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  /*max-width: 60%;*/
  /*transition: width .5s;*/
}

.Hand-icon:hover {
  width: 75px;
  height: 75px;
  /*max-width: 80%;*/
}

.Hand-icon--progress {
  composes: Hand-icon;
}
.Hand-icon--progress:hover {
  width: 50px;
  height: 50px;
}

.Hand-image {
  width: 100%;
  height: 100%;
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
  border-radius: 50px;
}

.Hand-image--rock {
  composes: Hand-image;
  background-image: url('../../../../../assets/rock.png');
}
.Hand-image--paper {
  composes: Hand-image;
  background-image: url('../../../../../assets/paper.png');
}
.Hand-image--scissors {
  composes: Hand-image;
  background-image: url('../../../../../assets/scissors.png');
  background-size: 70%;
}
.Hand-image--lizard {
  composes: Hand-image;
  background-image: url('../../../../../assets/lizard.png');
  background-size: 75%;
  background-position-x: 75%;
}
.Hand-image--spock {
  composes: Hand-image;
  background-image: url('../../../../../assets/spock.png');
  background-size: 65%;
}

.Hand-image--classicWaiting {
  composes: Hand-image;
  background-size: 65%;
  animation: classicWait .6s infinite;
}
.Hand-image--advancedWaiting {
  composes: Hand-image;
  background-size: 65%;
  animation: advancedWait 1s infinite;
}

@keyframes advancedWait {
    0%   {
      background-image: url('../../../../../assets/rock.png');
    }
    20%   {
      background-image: url('../../../../../assets/paper.png');
    }
    40%   {
      background-image: url('../../../../../assets/scissors.png');
    }
    60%   {
      background-image: url('../../../../../assets/lizard.png');
    }
    80%   {
      background-image: url('../../../../../assets/spock.png');
    }
}

@keyframes classicWait {
    0%   {
      background-image: url('../../../../../assets/rock.png');
    }
    33%   {
      background-image: url('../../../../../assets/paper.png');
    }
    66%   {
      background-image: url('../../../../../assets/scissors.png');
    }
}
