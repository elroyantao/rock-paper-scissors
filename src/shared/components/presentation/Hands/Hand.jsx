import React, { Component } from 'react'
import PropTypes from 'prop-types'

import styles from './Hand.css'

export default class Hand extends Component {
  static propTypes = {
    type: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string
    ]),
    isProgress: PropTypes.bool,
    onClick: PropTypes.func
  }

  static defaultProps = {
    onClick: () => {}
  }

  getButtonName(type) {
    let name
    switch (type) {
      case 0:
        name = 'rock'
        break
      case 1:
        name = 'spock'
        break
      case 2:
        name = 'paper'
        break
      case 3:
        name = 'lizard'
        break
      case 4:
        name = 'scissors'
        break
      default:
        name = type
    }
    return (
      <div className={styles[`Hand-image--${name}`]} />
    )
  }

  applyHand = () => {
    const { onClick, type } = this.props
    if (typeof type === 'string') return
    onClick(type)
  }

  render() {
    const { type, isProgress } = this.props
    return (
      <div className={styles[`Hand${isProgress ? '--progress' : ''}`]}>
        <div className={styles[`Hand-icon${isProgress ? '--progress' : ''}`]} onClick={this.applyHand}>
          {this.getButtonName(type)}
        </div>
      </div>
    )
  }
}
