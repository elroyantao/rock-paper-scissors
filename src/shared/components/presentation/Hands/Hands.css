.Hands {
  max-width: 100%;
  text-align: center;
}
.Hands-list {
  display: flex;
  justify-content: space-around;
  align-items: center;
  height: 80px;
}
