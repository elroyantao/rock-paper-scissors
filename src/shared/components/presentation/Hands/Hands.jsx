import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import Hand from './Hand'
import styles from './Hands.css'


@connect((state) => ({
  hands: state.level.hands,
  playerHand: state.hand.playerHand
}))
export default class Controls extends Component {
  static propTypes = {
    hands: PropTypes.array,
    playerHand: PropTypes.number,
    applyHand: PropTypes.func
  }

  render() {
    const { hands, playerHand, applyHand } = this.props
    if (!hands || !hands.length || playerHand !== null) return null
    return (
      <div className={styles.Hands}>
        <h3>Please play your hand</h3>
        <div className={styles['Hands-list']}>
          {
            hands.map((hand) =>
              <Hand type={hand} key={hand} onClick={applyHand} />
            )
          }
        </div>
      </div>
    )
  }
}
