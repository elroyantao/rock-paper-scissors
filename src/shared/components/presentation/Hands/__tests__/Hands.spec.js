import testComponentHelper from '../../../../lib/test-helper'
import Hands from '../Hands'


describe('<Hands /> ', () => {
  const initialState = {
    hands: [0, 2, 4],
    playerHand: null,
    applyHand: jest.fn()
  }
  const renderComponent = testComponentHelper(Hands.WrappedComponent)
  describe('@renders', () => {
    it('in default state', () => {
      expect(renderComponent(initialState).getTree())
        .toMatchSnapshot()
    })
    it('with playerHand set', () => {
      expect(renderComponent({
        ...initialState,
        playerHand: 3
      }).getTree())
        .toMatchSnapshot()
    })
    it('with hands as null', () => {
      expect(renderComponent({
        ...initialState,
        hands: []
      }).getTree())
        .toMatchSnapshot()
    })
  })
  describe('@events', () => {
    it('should call applyHand on click of hand', () => {
      const { wrapper, instance } = renderComponent(initialState)
      wrapper.find('Hand').first().simulate('click')
      expect(instance.props.applyHand).toHaveBeenCalledTimes(1)
    })
  })
})
