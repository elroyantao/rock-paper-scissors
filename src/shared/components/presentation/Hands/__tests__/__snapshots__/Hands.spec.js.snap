// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hands />  @renders in default state 1`] = `
<div
  className="Hands"
>
  <h3>
    Please play your hand
  </h3>
  <div
    className="Hands-list"
  >
    <Hand
      onClick={[Function]}
      type={0}
    />
    <Hand
      onClick={[Function]}
      type={2}
    />
    <Hand
      onClick={[Function]}
      type={4}
    />
  </div>
</div>
`;

exports[`<Hands />  @renders with hands as null 1`] = `null`;

exports[`<Hands />  @renders with playerHand set 1`] = `null`;
